next formatter

MonaMayrhofer opened the pull request:
# Objective

Fixes #315

This pull request implements a new formatter based on [dprint](https://dprint.dev/) to replace the existing "holdover" formatter.

# Status

- [X] Rudimentary support for all wgsl constructs
- [X] Format Ranges
- [ ] (Ongoing) Polish the formatting code.
- [X] Discuss which opinions the formatter should have, and nail that behavior down in the normative tests.
- [X] WESL

## Showcase

TBD

MonaMayrhofer committed:
Prepare testing infrastructure for new formatter

MonaMayrhofer committed:
Add some more smoke tests

MonaMayrhofer committed:
Rename the tests to smoke_tests because thats what they are

MonaMayrhofer committed:
Remove old formatter

MonaMayrhofer committed:
organize tests

MonaMayrhofer committed:
Prettyprint for code spacing newlines

MonaMayrhofer committed:
More testcases around module item spacing

MonaMayrhofer committed:
Update DesignNotes and Formatting.md

MonaMayrhofer committed:
Move out helpers

MonaMayrhofer committed:
Add error reporting to the formatter

MonaMayrhofer committed:
more exploratory work on fn signature layout

MonaMayrhofer committed:
integrate comment handling into fn signatures

MonaMayrhofer committed:
improve gen_spaced_lines

MonaMayrhofer committed:
Improve formatting structure

MonaMayrhofer committed:
Fix top level comments

BenjaminBrienen commented on crates/wgsl_formatter/src/format.rs:
```diff
@@ -195,27 +200,36 @@ fn gen_comments(
+        *last_item_was_space_or_newline = true;
+    } else {
+        //TODO Make this unrepresentable
+        unreachable!("Non comment entry found in comments Vec");
```
I wish rust had typescript types like `item: &SyntaxToken { kind: BlockComment | LineEndingComment },`

MonaMayrhofer replied:
Maybe in another life we get refinement types in rust... :,)

I think in the meantime an enum that stores the to_string-ed version of the comments suffices... or maybe I get some idea to avoid this problem in the first place

MonaMayrhofer committed:
Handle spacing via the PrintItemBuffer

MonaMayrhofer committed:
Split gen_* functions into files

MonaMayrhofer committed:
Rename PrintItemRequest

MonaMayrhofer committed:
Implement basic struct formatting

MonaMayrhofer committed:
work on struct/comment interop

MonaMayrhofer committed:
implements structs with attributes on members

MonaMayrhofer committed:
implement formatting compound statements

MonaMayrhofer committed:
implement formatting infix exprs

MonaMayrhofer committed:
implement long infix-expr line splitting

MonaMayrhofer committed:
Rename request_ shorthand methods that expect a separation to expect_

MonaMayrhofer committed:
work on formatting paren exprs

MonaMayrhofer committed:
add reasons to out of scope tests

MonaMayrhofer committed:
add prefix exprs

MonaMayrhofer committed:
Polishing tests for let and infix exprs

MonaMayrhofer committed:
cleanup tests

MonaMayrhofer committed:
basic function call

MonaMayrhofer committed:
index and field exprs

MonaMayrhofer committed:
var decls

MonaMayrhofer committed:
format basic while statements

MonaMayrhofer committed:
loop and continuing statements

MonaMayrhofer committed:
Merge high-changes into next-formatter

MonaMayrhofer committed:
break if statements

MonaMayrhofer committed:
clean up test warnings

MonaMayrhofer committed:
format return statement

MonaMayrhofer committed:
format simple if statements

MonaMayrhofer committed:
basic for statement formatting

MonaMayrhofer committed:
rework print_item_buffer request collapsing

MonaMayrhofer committed:
cleanup request_folder

MonaMayrhofer committed:
switch statement formatting

MonaMayrhofer committed:
switch statement default statement

MonaMayrhofer committed:
Reduce noise of comment tests

MonaMayrhofer committed:
begin moving unit-tests to check_comments

MonaMayrhofer committed:
Cleanup tests and add some new ones

MonaMayrhofer committed:
Add prefer-immediate-crash feature

MonaMayrhofer committed:
Move more tests to check_comments

MonaMayrhofer committed:
Remove parenthesis from while, switch, if

MonaMayrhofer committed:
assignment statement

MonaMayrhofer committed:
allow types for bindings

MonaMayrhofer committed:
increment decrement statement

MonaMayrhofer committed:
Fix non Ident-Expression in index and field access

MonaMayrhofer committed:
compound assignment statement

MonaMayrhofer committed:
const assert statements

MonaMayrhofer committed:
Add test for break if comments with parens

MonaMayrhofer committed:
merge ast-fixes

MonaMayrhofer committed:
fix compile after merge

MonaMayrhofer committed:
many cleanups

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
Add type formatting to let,const,var

MonaMayrhofer committed:
Tests for type exprs and more type handling in idents

MonaMayrhofer committed:
Add missing questionmark to parse_end's

MonaMayrhofer committed:
add unused lint and thusly fix a few bugs

MonaMayrhofer committed:
Fix const_assert statement tests

MonaMayrhofer committed:
Add tests

MonaMayrhofer committed:
Added failing tests for template elaborated function calls

MonaMayrhofer committed:
add global const and var statements

MonaMayrhofer committed:
Work on attribute formatting

MonaMayrhofer committed:
gen_comments takes a reference

MonaMayrhofer committed:
Add unit tests for attribute positions

MonaMayrhofer committed:
single attribute formatting

MonaMayrhofer committed:
worked on attribute positioning

MonaMayrhofer committed:
redo logic around lines and blankspaces

MonaMayrhofer committed:
update now working attribute_positions tests

MonaMayrhofer committed:
attributes on if statements

MonaMayrhofer committed:
attributes on switch statements

MonaMayrhofer committed:
attributes on loop statements

MonaMayrhofer committed:
attributes on for statements

MonaMayrhofer committed:
attributes on continuing statements

MonaMayrhofer committed:
attributes on while statements

MonaMayrhofer committed:
ignore everfailing tests

MonaMayrhofer committed:
override declarations

MonaMayrhofer committed:
top level code spacing follows compoundstatement

MonaMayrhofer committed:
Fix comment positioning after top level items

MonaMayrhofer committed:
Fix comments around empty/nonempty blocks

MonaMayrhofer committed:
fix comments around switch statements

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
fix build after merging main

MonaMayrhofer committed:
add some tests

MonaMayrhofer committed:
Add test for fn parameter spacing

MonaMayrhofer committed:
Fix formatting on semicolon after structs

MonaMayrhofer committed:
Fix formatting of non-literal indices

MonaMayrhofer committed:
polish struct member comment positioning

MonaMayrhofer committed:
clean up some tests

MonaMayrhofer committed:
polish logic around commets and spacing in fn parameters

MonaMayrhofer committed:
update smoke tests

MonaMayrhofer committed:
integrate wgsl-analyzer with the formatter

MonaMayrhofer committed:
begin gluing range formatting into lsp

stefnotch commented on crates/ide/src/formatting.rs:
```diff
@@ -8,21 +8,36 @@ use crate::RootDatabase;
+            NodeOrToken::Node(node) => node,
             NodeOrToken::Token(_) => return None,
         },
     };
```
I think we have to deal with the fact that the `node.text_range()` will be different from the `range` that is passed in.

Maybe the return type should be `Option<(TextRange, String)>`, where `TextRange` holds the original range that is getting replaced?

MonaMayrhofer committed:
range formatting in lsp

MonaMayrhofer committed:
fix smoke tests

MonaMayrhofer committed:
reorganize tests

MonaMayrhofer committed:
Add test for comment layout

MonaMayrhofer committed:
Grouped no-space-before-semicolon tests

MonaMayrhofer committed:
remove old request api calls

MonaMayrhofer committed:
fix some tests

MonaMayrhofer committed:
add test for parenthesis collapsing

MonaMayrhofer committed:
handling for multiline block comments

MonaMayrhofer committed:
Add test for index expression

MonaMayrhofer committed:
improve multiline group api

MonaMayrhofer committed:
start experimental MultilineGroup api

MonaMayrhofer committed:
use MultilineGroup everywhere

MonaMayrhofer committed:
remove separationrequest api

MonaMayrhofer committed:
rename reqest_request to request

MonaMayrhofer committed:
special handling for indent signals

MonaMayrhofer committed:
gen_import_statement stub

MonaMayrhofer committed:
simplify parser errors of the formatter

MonaMayrhofer committed:
small cleanups

MonaMayrhofer committed:
cleanup signals

MonaMayrhofer committed:
more little cleanups

MonaMayrhofer committed:
Polish api and testing around range formatting

MonaMayrhofer committed:
remove indoc

MonaMayrhofer committed:
experiment with multilinebuffer

MonaMayrhofer committed:
adapt tests to more sensible line breaking

MonaMayrhofer committed:
display file path when wgslfmt fails

MonaMayrhofer committed:
forward parser errors to wgslfmt

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
adapt cli

MonaMayrhofer committed:
adapt todo_verbatim_wesl to better handle multilines

MonaMayrhofer committed:
add tests

MonaMayrhofer committed:
fix new_line_group position within multiline_group

MonaMayrhofer committed:
More test

MonaMayrhofer committed:
more failing tests

MonaMayrhofer committed:
fix indentation for assignment statements

MonaMayrhofer committed:
adapt the config to what we actually use

MonaMayrhofer committed:
keep empty brackets together

MonaMayrhofer committed:
add some indentation tests

MonaMayrhofer committed:
fix indent with return statements

MonaMayrhofer committed:
work on attribute ordering tests

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
add tests

MonaMayrhofer committed:
add attribute ordering

MonaMayrhofer committed:
tests

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
diagnostic attribute

MonaMayrhofer committed:
enable and requires directives

MonaMayrhofer committed:
diagnostic attribute

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
gen_node via match

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
adapt to ast changes

MonaMayrhofer committed:
fix const attribute

MonaMayrhofer committed:
more diagnostic stuff

MonaMayrhofer committed:
Remove old code

MonaMayrhofer committed:
prepare import

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
fix attribute formatting

MonaMayrhofer committed:
work on imports

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
import line breaking

MonaMayrhofer committed:
path breaking

MonaMayrhofer committed:
skip tests that are not yet implemented

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
comments in imports

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
break contine discard into gen function

MonaMayrhofer committed:
break for parts into gen functions

MonaMayrhofer committed:
fix up imports

MonaMayrhofer committed:
remove outdated todos

MonaMayrhofer committed:
make check_comments ignore multilines

MonaMayrhofer committed:
remove redundant tests

MonaMayrhofer committed:
enable ignored test

MonaMayrhofer committed:
move statements into own mod

MonaMayrhofer committed:
split all statements into their own file

MonaMayrhofer committed:
move expressions into their own files

MonaMayrhofer committed:
pull statemetn/expr dispatchers into module file

MonaMayrhofer committed:
Rename expression files

More expressive that way, and we get rid of
gen_expression::gen_expression

also gen_expression could have more than just the gen_expression
function - it contains all the things related to the expression

MonaMayrhofer committed:
rename statement files

MonaMayrhofer committed:
rename format to generators

MonaMayrhofer committed:
move non generators out of generators module

MonaMayrhofer committed:
rename other generator files

MonaMayrhofer committed:
clean up design notes

MonaMayrhofer committed:
Remove Formatting.md

The normative test suite is now the source of truth.
Trying to specify the whole formatter in one md file
is likely a futile task, just because there are SO many
cases.

MonaMayrhofer committed:
debug assert linecomment newlines

MonaMayrhofer committed:
work on attribute formatting

MonaMayrhofer committed:
remove outdated todos

MonaMayrhofer committed:
add simple benchmark

BenjaminBrienen commented on crates/wgsl_formatter/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ serde = { workspace = true, optional = true }
 dprint-core = { version = "0.67.4", features = ["tracing"] }
 dprint-core-macros = "0.1.0"
 itertools = "0.14.0"
+thiserror = {version = "2.0.17"}
```
What's wrong with using `thiserror`? It's a great crate.

MonaMayrhofer committed:
remove old file

MonaMayrhofer committed:
Switch from BTreeMap to a simple bitmap

Making the code both cleaner and dropping the benchmark
from 1.9ms to 1.1ms on my machine. Sweet!!

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
remove default new function

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
simplify multi line group

MonaMayrhofer committed:
remove unnecessary to_string

MonaMayrhofer committed:
cleanups in request folder

MonaMayrhofer committed:
allow linebreaks in field statements

MonaMayrhofer committed:
add a few expression tests

MonaMayrhofer committed:
cleanup tests

MonaMayrhofer committed:
cleanup tests

MonaMayrhofer committed:
add test for breaking within fn parameter

MonaMayrhofer committed:
cleanup tests even more

MonaMayrhofer committed:
remove old todo

MonaMayrhofer committed:
Remove TODO

The api is fine as it is, no need to complicate things

MonaMayrhofer committed:
Remove TODO

Already Implemented

MonaMayrhofer committed:
Remove TODO

There is no way formatted will
escape to the user if there is an error.

MonaMayrhofer committed:
Remove TODO

There is no way that a malformed formatted will
escape to the user if the error is none.

MonaMayrhofer committed:
Cleanup of Attribute Code

MonaMayrhofer committed:
Remove TODO

Already done

MonaMayrhofer committed:
begin rework of cli

MonaMayrhofer committed:
json and text output for write format

MonaMayrhofer committed:
rework cli output

MonaMayrhofer committed:
add silent flag for stdin formatting

MonaMayrhofer committed:
remove old code

MonaMayrhofer committed:
basic touchups of readme

MonaMayrhofer committed:
Annotate and work through TODOs

MonaMayrhofer committed:
move logic out of request folder into request

MonaMayrhofer committed:
clarify request combination commutativity

MonaMayrhofer committed:
remove request_folder

MonaMayrhofer committed:
rename request_folder to spacing_request

MonaMayrhofer committed:
document spacing_request

MonaMayrhofer committed:
annotate and sqish TODOs

MonaMayrhofer committed:
annotate and sqish TODOs

MonaMayrhofer committed:
rename separated values

MonaMayrhofer committed:
add infallible comment api

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
change formatting of separated items

MonaMayrhofer committed:
change layout of comments in template lists

MonaMayrhofer committed:
absorb unnecessary comment handling into separated_items

MonaMayrhofer committed:
Remove outdated TODOs

MonaMayrhofer committed:
add Patterns.md

MonaMayrhofer committed:
Remove DesignNodes.md

MonaMayrhofer committed:
begin implementing switch collapse

MonaMayrhofer committed:
update tests to reflect default case formatting

MonaMayrhofer committed:
remove outdated todo

MonaMayrhofer committed:
implement debug assertions for multiline group

MonaMayrhofer committed:
phase out old .discourage api

MonaMayrhofer committed:
cargo fix

MonaMayrhofer committed:
remove outdated todo

MonaMayrhofer committed:
move context policies from params into module

MonaMayrhofer committed:
prettify mat3x3

MonaMayrhofer committed:
move vec3f expansion to post 1.0

MonaMayrhofer committed:
cleanups

BenjaminBrienen commented on crates/wgsl_formatter/Cargo.toml:
```diff
@@ -1,24 +1,40 @@
-rust-version.workspace = true
-description = "RPC Api for the `proc-macro-srv` crate of wgsl-analyzer."
 repository.workspace = true
+description = "RPC Api for the `proc-macro-srv` crate of wgsl-analyzer."
```
Huh?

MonaMayrhofer replied:
OOpsie, How long has that been there :O

MonaMayrhofer committed:
merge main

MonaMayrhofer committed:
TODO FIXES

MonaMayrhofer committed:
Support new syntax

MonaMayrhofer committed:
fix one test that should induce a rats tail of fixes

MonaMayrhofer committed:
work on new trivia implementation

MonaMayrhofer committed:
Make idempotence check error clearer.

MonaMayrhofer committed:
move source_file to new trivia

MonaMayrhofer committed:
work on new trivia implementation

MonaMayrhofer committed:
fix attr spacing

MonaMayrhofer committed:
fix @attribute being a reserved keyword

MonaMayrhofer committed:
move switch to new trivia

MonaMayrhofer committed:
fix attributes in loop and continuing

MonaMayrhofer committed:
move fn parameters to new trivia

MonaMayrhofer committed:
parse_node_with_trivia_until

MonaMayrhofer committed:
fix struct with trivia_filter

MonaMayrhofer committed:
attributes always belong to the next item

MonaMayrhofer committed:
move everything to new IgnoreBlankspace

MonaMayrhofer committed:
rename parse_node_with_trivia_until

MonaMayrhofer committed:
assignment_statement

MonaMayrhofer committed:
grep replace parse_token

MonaMayrhofer committed:
grep replace parse_token_optional

MonaMayrhofer committed:
break if

MonaMayrhofer committed:
only_if_kind filter

MonaMayrhofer committed:
move easy cases to parse_node_with

MonaMayrhofer committed:
Make syntax_iter a function

MonaMayrhofer committed:
move imports to parse_node_with

MonaMayrhofer committed:
move incdec to parse_node_with

MonaMayrhofer committed:
move var let const to parse_node_with

MonaMayrhofer committed:
move diagnostic directive to parse_node_with

MonaMayrhofer committed:
function call statement

MonaMayrhofer committed:
switch case selectors

MonaMayrhofer committed:
import super relative

MonaMayrhofer committed:
path

MonaMayrhofer committed:
struct declaration

MonaMayrhofer committed:
types

MonaMayrhofer committed:
move attributes to new trivia handing

MonaMayrhofer committed:
implement warning for unclosed syntax iter

MonaMayrhofer committed:
cleanups

MonaMayrhofer committed:
cleanups

BenjaminBrienen commented:
🙏
